### PR TITLE
JAV-501 Public key authentication between consumer and provider

### PR DESCRIPTION
**Pull request by coolhongluo**
As a developer, I need to have authentication in service provider to protect the microservice from unauthorized access

### CODE REVIEW

File: foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAUtils.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+  private static KeyFactory kf = null;
+  
+  static {
+    
+    try
+    {
+      kf = KeyFactory.getInstance(RSA_ALG);
+    }catch(NoSuchAlgorithmException e)
+    {
+      LOGGER.error("init keyfactory error");
```
Comment (liubao68) [inline foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAUtils.java]: Code not well formated

File: foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAUtils.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    }
+  }
+
+  /**
+   * if has performance problem ,change Signature to ThreadLocal instance 
+   */
+  public static String sign(String content, PrivateKey privateKey)
+      throws NoSuchAlgorithmException, InvalidKeySpecException, SignatureException, InvalidKeyException {
+    Signature signature = Signature.getInstance(SIGN_ALG);
+    signature.initSign(privateKey);
```
Comment (liubao68) [inline foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAUtils.java]: Public and private key may not change in every call, for performance we can cache the init values. It is fine now and need more test data to evaluate

File: handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/RSAAuthenticationToken.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    return serviceId;
+  }
+
+  public void setServiceId(String serviceId) {
+    this.serviceId = serviceId;
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    if (null == obj || !(obj instanceof RSAAuthenticationToken)) {
```
Comment (liubao68) [inline handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/RSAAuthenticationToken.java]: redundant expression

File: handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/consumer/ConsumerAuthHandler.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+ */
+public class ConsumerAuthHandler implements Handler {
+
+  private RSAConsumerTokenManager athenticationTokenManager = new RSAConsumerTokenManager();
+
+  @Override
+  public void handle(Invocation invocation, AsyncResponse asyncResp) throws Exception {
+
+    Optional<String> token = Optional.ofNullable(athenticationTokenManager.getToken());
+    if(!token.isPresent())
```
Comment (liubao68) [inline handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/consumer/ConsumerAuthHandler.java]: I think, we can delete all these instructions.

File: handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/consumer/RSAConsumerTokenManager.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    long generateTime = System.currentTimeMillis();
+    try {
+      String plain = String.format("%s@%s@%s@%s", instanceId, serviceId, generateTime, randomCode);
+      String sign = RSAUtils.sign(plain, privateKey);
+      token = RSAAuthenticationToken.fromStr(String.format("%s@%s", plain, sign));
+      return token.format();
+    } catch (InvalidKeyException | NoSuchAlgorithmException | InvalidKeySpecException | SignatureException e) {
+      logger.error("create token error", e);
+      throw new Error("create token error");
+    }
```
Comment (liubao68) [inline handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/consumer/RSAConsumerTokenManager.java]: lock is not properly released in some situation. I think we do not need lock at all at this part. e.g. we just add a one time task to do the creation of the key or we can just lock the procedure of creating the key.

File: handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/provider/RSAProviderTokenManager.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+  private boolean tokenExprired(RSAAuthenticationToken rsaToken) {
+    long generateTime = rsaToken.getGenerateTime();
+    long expired = generateTime + RSAAuthenticationToken.TOKEN_ACTIVE_TIME + 15 * 60 * 1000;
+    long now = System.currentTimeMillis();
+    return now > expired;
+  }
+
+  private String getPublicKey(String instanceId, String serviceId) {
+    Optional<MicroserviceInstance> instances = Optional
```
Comment (liubao68) [inline handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/provider/RSAProviderTokenManager.java]: This is just a null check, why need using Optional？

File: service-registry/src/main/java/io/servicecomb/serviceregistry/cache/MicroserviceInstanceCache.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+ *
+ */
+public class MicroserviceInstanceCache {
+
+	private static final Logger logger = LoggerFactory.getLogger(MicroserviceInstanceCache.class);
+
+	private static Cache<String, MicroserviceInstance> instances = CacheBuilder.newBuilder().maximumSize(1000)
+			.expireAfterAccess(30, TimeUnit.MINUTES).build();
+
+	public static MicroserviceInstance getOrCreate(String serviceId, String instanceId) {
```
Comment (liubao68) [inline service-registry/src/main/java/io/servicecomb/serviceregistry/cache/MicroserviceInstanceCache.java]: do you use the code template under $HOME/etc folder?

File: foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAKeyPairEntry.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
+package io.servicecomb.foundation.common.utils;
```
Comment (liubao68) [inline foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAKeyPairEntry.java]: copy right

Comment (coolhongluo) [reply foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAKeyPairEntry.java]: accept

File: service-registry/src/main/java/io/servicecomb/serviceregistry/cache/MicroserviceInstanceCache.java (hunk 7 above)
Comment (coolhongluo) [reply service-registry/src/main/java/io/servicecomb/serviceregistry/cache/MicroserviceInstanceCache.java]: accept ; forget to format

File: handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/provider/RSAProviderTokenManager.java (hunk 6 above)
Comment (coolhongluo) [reply handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/provider/RSAProviderTokenManager.java]: accept

File: handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/consumer/ConsumerAuthHandler.java (hunk 4 above)
Comment (coolhongluo) [reply handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/consumer/ConsumerAuthHandler.java]: maybe,but i think can reduce send request to server if do it

File: handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/RSAAuthenticationToken.java (hunk 3 above)
Comment (coolhongluo) [reply handlers/handler-publickey-auth/src/main/java/io/servicecomb/authentication/RSAAuthenticationToken.java]: accept

File: foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAUtils.java (hunk 1 above)
Comment (coolhongluo) [reply foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAUtils.java]: new code ，forget to format

File: foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAUtils.java (hunk 2 above)
Comment (coolhongluo) [reply foundations/foundation-common/src/main/java/io/servicecomb/foundation/common/utils/RSAUtils.java]: this is a good idea,but i think the RSAUtils is a common function;
if have performance  problem ,i will refactor it

**Pull request merged**